File: ios/OpenClawConsole/OpenClawConsole/Models/GatewayPairing.swift
// Models/GatewayPairing.swift
// OpenClaw Work Console
// Parses QR/deep-link gateway pairing payloads.

import Foundation

struct GatewayPairing: Equatable {
    let name: String
    let baseURL: String
    let token: String

    enum ParseError: LocalizedError {
        case unsupportedFormat
        case missingField(String)
        case invalidURL

        var errorDescription: String? {
            switch self {
            case .unsupportedFormat:
                return "Paste an OpenClaw pairing link or pairing JSON."
            case .missingField(let field):
                return "Pairing code is missing \(field)."
            case .invalidURL:
                return "Pairing code contains an invalid gateway URL."
            }
        }
    }

    static func parse(_ rawValue: String) throws -> GatewayPairing {
        let raw = rawValue.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !raw.isEmpty else { throw ParseError.unsupportedFormat }

        if raw.hasPrefix("{") {
            return try parseJSON(raw)
        }

        guard let url = URL(string: raw),
              let components = URLComponents(url: url, resolvingAgainstBaseURL: false) else {
            throw ParseError.unsupportedFormat
        }

        let isPairingLink = url.scheme == "openclaw" && url.host == "pair"
        guard isPairingLink else { throw ParseError.unsupportedFormat }

        let items = Dictionary(uniqueKeysWithValues: (components.queryItems ?? []).map { ($0.name, $0.value ?? "") })
        return try build(
            name: items["name"],
            baseURL: items["base_url"] ?? items["baseUrl"] ?? items["baseURL"],
            token: items["token"]
        )
    }

    private static func parseJSON(_ raw: String) throws -> GatewayPairing {
        guard let data = raw.data(using: .utf8) else { throw ParseError.unsupportedFormat }
        let payload = try JSONDecoder().decode(GatewayPairingJSONPayload.self, from: data)
        guard payload.type == "openclaw.gateway.pairing.v1" else {
            throw ParseError.unsupportedFormat
        }
        return try build(name: payload.name, baseURL: payload.baseURL, token: payload.token)
    }

    private static func build(name: String?, baseURL: String?, token: String?) throws -> GatewayPairing {
        let cleanedName = (name ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        let cleanedURL = (baseURL ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
            .trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        let cleanedToken = (token ?? "").trimmingCharacters(in: .whitespacesAndNewlines)

        guard !cleanedURL.isEmpty else { throw ParseError.missingField("base_url") }
        guard !cleanedToken.isEmpty else { throw ParseError.missingField("token") }
        guard cleanedURL.hasPrefix("https://") || cleanedURL.hasPrefix("http://") else {
            throw ParseError.invalidURL
        }

        return GatewayPairing(
            name: cleanedName.isEmpty ? "OpenClaw Gateway" : cleanedName,
            baseURL: cleanedURL,
            token: cleanedToken
        )
    }

}

private struct GatewayPairingJSONPayload: Decodable {
    let type: String
    let name: String?
    let baseURL: String?
    let token: String?

    enum CodingKeys: String, CodingKey {
        case type
        case name
        case baseURL = "base_url"
        case token
    }
}

File: android/app/src/main/java/com/openclaw/console/data/model/GatewayPairing.kt
package com.openclaw.console.data.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.net.URI
import java.net.URLDecoder
import java.nio.charset.StandardCharsets
import java.util.Base64

data class GatewayPairing(
    val name: String,
    val baseUrl: String,
    val token: String
) {
    companion object {
        private val json = Json { ignoreUnknownKeys = true }

        fun parse(rawValue: String): Result<GatewayPairing> = runCatching {
            val raw = rawValue.trim()
            require(raw.isNotEmpty()) { "Paste an OpenClaw pairing link or pairing JSON." }

            if (raw.startsWith("{")) {
                return@runCatching parseJson(raw)
            }

            parseSetupCode(raw)?.let { return@runCatching it }

            val uri = URI(raw)
            if (uri.scheme == "http" || uri.scheme == "https" || uri.scheme == "ws" || uri.scheme == "wss") {
                return@runCatching parseGatewayUrl(uri)
            }

            require(uri.scheme == "openclaw" && uri.host == "pair") {
                "Paste an OpenClaw pairing link or pairing JSON."
            }
            val query = parseQuery(uri.rawQuery.orEmpty())

            build(
                name = query["name"],
                baseUrl = query["base_url"] ?: query["baseUrl"] ?: query["baseURL"],
                token = query["token"] ?: query["tkn"]
            )
        }

        private fun parseJson(raw: String): GatewayPairing {
            val payload = json.decodeFromString<PairingJsonPayload>(raw)
            require(payload.type == "openclaw.gateway.pairing.v1") {
                "Paste an OpenClaw pairing link or pairing JSON."
            }
            return build(payload.name, payload.baseUrl ?: payload.url, payload.bootstrapToken ?: payload.token)
        }

        private fun parseSetupCode(raw: String): GatewayPairing? {
            val decoded = decodeSetupCode(raw) ?: return null
            val payload = json.decodeFromString<SetupCodePayload>(decoded)
            return build(
                name = payload.name,
                baseUrl = payload.baseUrl ?: payload.url,
                token = payload.bootstrapToken ?: payload.token
            )
        }

        private fun parseGatewayUrl(uri: URI): GatewayPairing {
            val query = parseQuery(uri.rawQuery.orEmpty())
            val token = query["bootstrapToken"] ?: query["token"] ?: query["tkn"]
            val path = uri.rawPath.orEmpty().removeSuffix("/api/health").removeSuffix("/")
            val port = if (uri.port == -1) "" else ":${uri.port}"
            val scheme = when (uri.scheme) {
                "wss" -> "https"
                "ws" -> "http"
                else -> uri.scheme
            }
            val baseUrl = "$scheme://${uri.host}$port$path"
            val name = uri.host?.takeIf { it.isNotBlank() }?.let { "OpenClaw $it" }

            return build(name = name, baseUrl = baseUrl, token = token)
        }

        private fun decodeSetupCode(raw: String): String? {
            if (raw.any { it.isWhitespace() }) return null
            return runCatching {
                String(Base64.getUrlDecoder().decode(raw), StandardCharsets.UTF_8)
            }.getOrNull()?.takeIf { it.trimStart().startsWith("{") }
        }

        private fun build(name: String?, baseUrl: String?, token: String?): GatewayPairing {
            val cleanedName = name.orEmpty().trim().ifBlank { "OpenClaw Gateway" }
            val cleanedBaseUrl = baseUrl.orEmpty().trim().trimEnd('/').normalizeGatewayUrl()
            val cleanedToken = token.orEmpty().trim()

            require(cleanedBaseUrl.isNotBlank()) { "Pairing code is missing base_url." }
            require(cleanedToken.isNotBlank()) { "Pairing code is missing token." }
            require(cleanedBaseUrl.startsWith("https://") || cleanedBaseUrl.startsWith("http://")) {
                "Pairing code contains an invalid gateway URL."
            }

            return GatewayPairing(
                name = cleanedName,
                baseUrl = cleanedBaseUrl,
                token = cleanedToken
            )
        }

        private fun parseQuery(rawQuery: String): Map<String, String> {
            if (rawQuery.isBlank()) return emptyMap()

            return rawQuery.split('&')
                .mapNotNull { part ->
                    val pieces = part.split('=', limit = 2)
                    val key = decode(pieces.getOrNull(0).orEmpty())
                    if (key.isBlank()) return@mapNotNull null
                    key to decode(pieces.getOrNull(1).orEmpty())
                }
                .toMap()
        }

        private fun decode(value: String): String =
            URLDecoder.decode(value, StandardCharsets.UTF_8.name())

        private fun String.normalizeGatewayUrl(): String = when {
            startsWith("wss://") -> replaceFirst("wss://", "https://")
            startsWith("ws://") -> replaceFirst("ws://", "http://")
            else -> this
        }
    }
}

@Serializable
private data class PairingJsonPayload(
    val type: String,
    val name: String? = null,
    @SerialName("base_url") val baseUrl: String? = null,
    val url: String? = null,
    val token: String? = null,
    val bootstrapToken: String? = null
)

@Serializable
private data class SetupCodePayload(
    val name: String? = null,
    @SerialName("base_url") val baseUrl: String? = null,
    val url: String? = null,
    val token: String? = null,
    val bootstrapToken: String? = null
)
